### ISSUE TITLE
Refactor the validation functions in DCA to incorporate changes from schematic 

### ISSUE THREAD

**Issue by rrchai** [Sage-Bionetworks/data_curator#309]
- Incorporate backend updates from https://github.com/Sage-Bionetworks/schematic/pull/614
---

- [ ] Ensure the updated validateModelManifest() work as expected
 refactor a validation function in DCA to cope the changes: e.g. validateModelManifest() now return two things: "error" and "warning". I think we just need first element (error) of output list in DCA
- [ ] Test cross-manifest rules on HTAN data:  add this rule to HTAN Parent Biospecimen ID: matchAtLeastOne scRNA-seqLevel1.HTANParentBiospecimenID Biospecimen.HTANBiospecimenID
- [ ] When it's a wrong schema error, I notice validateModelManifest() also return which column is missing: e.g "'Filename' is a required property". Not sure if it is a new feature, but I think we could tell users which columns they are missing to improve our errors for wrong schema .

**Comment by rrchai** [comment]
related to https://github.com/Sage-Bionetworks/schematic/issues/683:

>One thing to note for DCA:
Each row represents each invalid value with error in the UI error table, so "matchExactlyOne" might not be fit for the error table since entire column/attribute is invalid. Thus, we might consider "matchExactlyOne" errors as a separated error type, I could just highlight the invalid source attribute and throw its errors.

**Comment by rrchai** [comment]
## Editting:
I have run some tests on the cross-manifest testing (MatchAtLeast) using the latest schematic:
- [testing manifest](https://www.synapse.org/#!Synapse:syn29296703)
- [testing data model](https://github.com/ncihtan/data-models/blob/feature-cross-validation/HTAN.model.jsonld)
  - there is only one cross-manifest rule: 
  ```"matchAtLeastOne scRNA-seqLevel1.HTANParentBiospecimenID Biospecimen.HTANBiospecimenID"```

Here is my benchmarking result for running time (by running validation 5 times with ^ testing files using [hypefine](https://github.com/sharkdp/hyperfine):
Using the schematic before []() 700 is merged: 
```console
hyperfine --warmup 1 --runs 5 --ignore-failure 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1'
```

Using the latest schematic:
```console
hyperfine --warmup 1 --runs 5 --ignore-failure 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1'
```

This is my downloading speed:
```console
Testing download speed................................................................................
Download: 8.13 Mbit/s
Testing upload speed......................................................................................................
Upload: 11.04 Mbit/s
```

**Comment by milen-sage** [comment]
@rrchai this is exactly the kind of benchmarking we need more of! My take aways from it:
- we'll need project scoping - and repeat the same benchmark with it
- very likely, we'll still need filtering based on component annotation - that would require adding the component annotation to HTAN dataset manifests retrospectively 

@GiaJordan when you're back, what are your thoughts on project scoping readiness for testing?

 @ychae what do you think about who'd be best positioned to work on HTAN's retroactive component annotations of manifests?

**Comment by GiaJordan** [comment]
@rrchai 
In the development of this I noticed that when there were multiple manifests for a dataset, it would just take the first one and this wouldn't always be desired, so I added logic there for selection. It shouldn't affect datasets with only one manifest though.
The manifests currently on synapse will also need to have component annotations added either manually or by re-uploading for the expedited gathering to work, otherwise it will revert back to the old behavior.

@milen-sage 
Project scoping is ready for testing now; I'm planning on merging [#701](https://github.com/Sage-Bionetworks/schematic/pull/701) on Monday

**Comment by ychae** [comment]
> @ychae what do you think about who'd be best positioned to work on HTAN's retroactive component annotations of manifests?

@milen-sage I think the data liaisons might be able to help with this task -- ideally each liaison would add the component annotation for their centers but I'll need to float this idea by Ashley to see how keen the liaisons would be to take this on.

**Comment by rrchai** [comment]
@GiaJordan @milen-sage By validating on only HTAN center A using the latest schematic (merged [701](https://github.com/Sage-Bionetworks/schematic/pull/701), the speed is amazing now.

<details>
<summary>validation results</summary>

```python
error: Manifest syn27331282 does not contain the value HTA111111_121324 from row 2 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn27331282 does not contain the value sdsadasd from row 3 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn27331282 does not contain the value HTA3_8001_1002 from row 4 of the attribute HTANParentBiospecimenID in the source manifest.
```
</details>

Since there is no flag for project_scope using cli, I replace default value of [project_scope](https://github.com/Sage-Bionetworks/schematic/blob/14aff6e0a1765d443d74a7f903757a2f229e57a0/schematic/models/metadata.py#L189) with centerA's synId for testing.
 
```console
hyperfine --warmup 1 --runs 5 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1'
  Time (mean ± σ):     18.844 s ±  5.422 s    [User: 4.838 s, System: 0.672 s]
  Range (min … max):   11.972 s … 25.155 s    5 runs
```

**Comment by GiaJordan** [comment]
@rrchai I'm not sure why there's different behavior for using the flag and chaning the default value. Does the master_fileview for both contain the CenterA project?

The warning for the unmet access restrictions is expected, though I suppose displaying a message that a censored version of the manifest will be downloaded instead would be helpful.

Very happy to hear about the times!

**Comment by rrchai** [comment]
@GiaJordan ah, you are right. I forgot to use the HTAN's fileview. I will re-test it.

**Comment by milen-sage** [comment]
>  I forgot to use the HTAN's fileview. I will re-test it.

This might explain the speed up too. HTAN's fileview is larger.

**Comment by rrchai** [comment]
## Test HTAN fileview

<details>
<summary>validation results</summary>

```python
error: Manifest syn27116926 does not contain the value HTA111111_121324 from row 2 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn27116926 does not contain the value sdsadasd from row 3 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn27116926 does not contain the value HTA3_8001_1002 from row 4 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn30560343 does not contain the value HTA111111_121324 from row 2 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn30560343 does not contain the value sdsadasd from row 3 of the attribute HTANParentBiospecimenID in the source manifest.
error: Manifest syn30560343 does not contain the value HTA3_8001_1002 from row 4 of the attribute HTANParentBiospecimenID in the source manifest.
```
</details>

```console
hyperfine --warmup 1 --runs 5 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1 -ps syn20977135'
  Time (mean ± σ):     120.861 s ± 66.810 s    [User: 16.036 s, System: 1.864 s]
  Range (min … max):   54.659 s … 215.867 s    5 runs
```

The speed is definitely faster by using project scope. I think 2 min averagely for me is okay to use and test on one/two cross-manifests validation. I notice the most time it takes on my end is querying the fileview table?  
```
Downloading  [####################]100.00%   74.9MB/74.9MB (538.7kB/s) SYNAPSE_TABLE_QUERY_95623552.csv.synapse_download_95623552 Done...
``` 
Since the DCA has already stored the synapseStorage object at the beginning. @milen-sage I wonder if it possible to use the synapseStorage object directly as input to avoid downloading it again? I still can proceed to implement it in the DCA without this enhancement.

**Comment by rrchai** [comment]
@GiaJordan I tested the [develop-crossM-union-rule](https://github.com/Sage-Bionetworks/schematic/tree/develop-crossM-union-rule) branch:

1. using `"matchAtLeastOne Biospecimen.HTANBiospecimenID set"`:

It looks like it is validating individual values?

```console
warning: Value(s) ['HTA111111_121324'] from row(s) [2] of the attribute htanparentbiospecimenid in the source manifest are missing. Manifest(s) ['syn27116926', 'syn30560343'] are missing the value(s).
Your manifest has been validated successfully. There are no errors in your manifest, and it can be submitted without any modifications.
```

2. using `"matchAtLeastOne Biospecimen.HTANBiospecimenID value"`:

```console
warning: Value(s) ['HTA111111_121324'] from row(s) [2] of the attribute htanparentbiospecimenid in the source manifest are missing. Manifest(s) ['syn27116926', 'syn30560343'] are missing the value(s).
Your manifest has been validated successfully. There are no errors in your manifest, and it can be submitted without any modifications.
```

FYI: currently, the DCA will only extract the error list (4 elements) from validation results and reformat it into a table:
Row | Column | Value | Error |
--- | --- | --- | --- 
 | 1st element| 2rd element | 4th element | 3rd element (anything start with "not")

Warning could also be used as long as it has the same format of errors. I found the almost all errors from schematic contains "not". To simplify the UI error message, only characters after "not" (including "not") are used. So the error message (3rd element of error output list) prefers to contain "not".

**Comment by milen-sage** [comment]
@rrchai yes, let's give reusing the fileview object a try! If this avoids the download time, it would be worth it for sure. 

In this case it sounds like there are two items:

1. integration of the cross-manifest rule in the DCA and the dashboard
2. fileview object reuse in the dashboard

If the implementation of 2 doesn't affect 1 too much, you can start with 1 and then move on to 2. 

On the other hand, if 2 will affect significantly how 1 is implemented, proceeding with 2 and testing it first would make more sense.

My intuition is that 2 shouldn't affect too much 1. What do you think?

**Comment by GiaJordan** [comment]
@rrchai Thanks for the feedback. It will validate on the set or value level, but with either case it'll display which values are causing the invalidity. And while they've been switched to warnings, the format is still the same.

In response to the query being the limiting step currently, I added the project scope to the table query if one is provided. That should help to speed things up more. This change is still the cross manifest branch, so you can pull the latest version to try. 
cc @milen-sage

**Comment by rrchai** [comment]
okay,  my internet issue has been fixed and the downloading speed is back to normal:

```console
Testing download speed................................................................................
Download: 170.74 Mbit/s
Testing upload speed......................................................................................................
Upload: 9.92 Mbit/s
```
I retested the latest develop branch and downloading fileview table is not as slow as before with ^ speed 😄 :

## Using the latest develop branch:
```console
hyperfine --warmup 1 --runs 10 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1 -ps syn20977135'
  Time (mean ± σ):     42.877 s ±  5.390 s    [User: 15.350 s, System: 1.220 s]
  Range (min … max):   37.302 s … 51.442 s    10 runs
```

## Using  [develop-crossM-union-rule](https://github.com/Sage-Bionetworks/schematic/tree/develop-crossM-union-rule) branch:
```console
hyperfine --warmup 1 --runs 10 'schematic model -c schematic_config.yml validate -mp ~/Downloads/htan/test-cross-link.csv -dt ScRNA-seqLevel1 -ps syn20977135'
  Time (mean ± σ):     23.530 s ±  2.253 s    [User: 5.753 s, System: 0.740 s]
  Range (min … max):   20.351 s … 26.741 s    10 runs
```

--- 
@GiaJordan Thank you for the improvement. The new querying looks like that it reduces time significantly from 40ish to 20ish 🚀 ! 

> My intuition is that 2 shouldn't affect too much 1. What do you think?
Based on the benchmarking results, I think downloading fileview object will affect the overall speed of validation. Since the @GiaJordan has already improve the query process and ~20s for one cross-manifest validation is good to me, I could start to implement the cross-manifest rules in the app now. 

In short, by testing with one cross-manifest rule with my internet, the faster your downloading speed is and less projects/data to query/download, the faster the cross-manifest validation will be.

**Comment by GiaJordan** [comment]
@rrchai Thanks for the benchmarking and updates!
In the farther future, we'll likely experiment with converting the synapse storage object to a sentinel object that can be reused within one validation run. This would expedite the validation further in cases where multiple attributes use the cross manifest validation rule.

**Comment by milen-sage** [comment]
Looks like we just need to review a [PR](https://github.com/Sage-Bionetworks/data_curator/pull/347) and this will be ready to close?

**Comment by GiaJordan** [comment]
@milen-sage can we then merge [#751](https://github.com/Sage-Bionetworks/schematic/pull/751) as well?
